{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-30T07:32:46.676379Z",
     "start_time": "2018-11-30T07:32:46.396745Z"
    }
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Author Hanqing Liu\n",
    "\n",
    "This file contain functions for using Primer3\n",
    "\n",
    "Input:\n",
    "- A bed file indicates targeted region/sites\n",
    "- Genome fasta file corresponding to the bed file\n",
    "- Config file for primer 3\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import subprocess\n",
    "import collections\n",
    "import multiprocessing\n",
    "from functools import partial\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio import SearchIO\n",
    "\n",
    "\n",
    "def _read_bed(file_path):\n",
    "    bed_df = pd.read_table(file_path,\n",
    "                           index_col='region_id', header=None, comment='#',\n",
    "                           names=['seq_name', 'start', 'end', 'region_id'])\n",
    "    return bed_df\n",
    "\n",
    "\n",
    "def _read_fasta_fai(fai_path):\n",
    "    # read samtools faidx format\n",
    "    fai_df = pd.read_table(fai_path,\n",
    "                           index_col=0, header=None,\n",
    "                           names=['seq_name', 'length', 'start_at',\n",
    "                                  'line_seq_length', 'line_total_length'])\n",
    "    return fai_df\n",
    "\n",
    "\n",
    "def _query_genome(fasta_path, fai_df,\n",
    "                  seq_name, region_start, region_end,\n",
    "                  left_expand, right_expand, primer_name):\n",
    "    # check fai, get position\n",
    "    if seq_name not in fai_df.index:\n",
    "        raise KeyError(f'{seq_name} not in the faidx file of genome fasta {fasta_path}')\n",
    "    if region_end < region_start:\n",
    "        raise ValueError(\n",
    "            f'Region end {region_end} < Region Start {region_start} at primer {primer_name}, check your input')\n",
    "\n",
    "    # calculate length\n",
    "    seq_start_pos = fai_df.loc[seq_name, 'start_at']\n",
    "    # in case the region start is close to ref sequence start\n",
    "    chrom_query_start = max(0, seq_start_pos + region_start - left_expand)\n",
    "    if chrom_query_start == 0:\n",
    "        real_left_expand = seq_start_pos\n",
    "    else:\n",
    "        real_left_expand = left_expand\n",
    "    chrom_query_length = real_left_expand + (region_end - region_start) + right_expand\n",
    "\n",
    "    # check some error\n",
    "    if chrom_query_length > 999999:\n",
    "        raise ValueError(f'At primer {primer_name}, do you really want to design primer with lenght > 999999?')\n",
    "    if fasta_path.endswith('.gz'):\n",
    "        raise NotImplementedError(\n",
    "            'Genome fasta is gziped, query sequence from gzip file could be super slow, '\n",
    "            'unzip it should be much faster.')\n",
    "\n",
    "    # get sequence\n",
    "    with open(fasta_path) as f:\n",
    "        f.seek(chrom_query_start)\n",
    "        sequence_context = ''\n",
    "        for line in f:\n",
    "            if line[0] == '>':  # read to next seq\n",
    "                break\n",
    "            elif len(sequence_context) >= chrom_query_length:  # read enough length\n",
    "                break\n",
    "            else:\n",
    "                sequence_context += line.strip()\n",
    "        query_sequence = sequence_context[:min(chrom_query_length, len(sequence_context))]\n",
    "    if len(sequence_context) >= chrom_query_length:\n",
    "        real_right_expand = right_expand\n",
    "    else:\n",
    "        real_right_expand = len(sequence_context) - real_left_expand - (region_end - region_start)\n",
    "    if real_right_expand < 0:\n",
    "        raise ValueError(f'At primer {primer_name}, Region end is outside reference sequence.')\n",
    "\n",
    "    query_result = pd.Series({  # following primer3 tag names\n",
    "        'SEQUENCE_NAME': primer_name,\n",
    "        'SEQUENCE_TEMPLATE': query_sequence.upper(),\n",
    "        'SEQUENCE_TARGET': f'{real_left_expand},{region_end - region_start}'\n",
    "    })\n",
    "    return query_result\n",
    "\n",
    "\n",
    "def _run_primer3(primer_template_df, setting_dict, cpu):\n",
    "    pool = multiprocessing.Pool(cpu)\n",
    "    primer3_runner = partial(subprocess.run,\n",
    "                             args=['primer3_core'],\n",
    "                             stdout=subprocess.PIPE,\n",
    "                             encoding='utf8',\n",
    "                             check=False)\n",
    "    results = []\n",
    "    for primer_name, record in primer_template_df.iterrows():\n",
    "        # modify setting_dict, prepare primer specific input\n",
    "        row_input_dict = setting_dict.copy()\n",
    "        for k, v in record.iteritems():\n",
    "            if k != '':\n",
    "                row_input_dict[k] = v\n",
    "        row_input_dict['SEQUENCE_NAME'] = primer_name\n",
    "        primer3_input = ''\n",
    "\n",
    "        for k, v in row_input_dict.items():\n",
    "            if k != '':\n",
    "                primer3_input += f'{k}={v}\\n'\n",
    "        primer3_input += '=\\n'\n",
    "        # run\n",
    "        result = pool.apply_async(primer3_runner,\n",
    "                                  kwds=dict(input=primer3_input))\n",
    "        results.append(result)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "\n",
    "    primer_stats = []\n",
    "    primers = []\n",
    "    for result in results:\n",
    "        primer3_return = result.get()\n",
    "        if primer3_return.returncode != 0:\n",
    "            print(primer3_return)\n",
    "            return\n",
    "\n",
    "        primer3_out = primer3_return.stdout\n",
    "        primer_stat_df, primer_df = _parse_primer3_result(primer3_out)\n",
    "        primer_stats.append(primer_stat_df)\n",
    "        primers.append(primer_df)\n",
    "    total_primer_stat_df = pd.concat(primer_stats, sort=True)\n",
    "    total_primer_df = pd.concat(primers, sort=True)\n",
    "\n",
    "    return primer_template_df, total_primer_stat_df, total_primer_df\n",
    "\n",
    "\n",
    "def _parse_primer3_result(primer3_out):\n",
    "    # split stat part and primer part\n",
    "    stat_dict = {}\n",
    "    primer_dict = {}\n",
    "    primer_info_start = False\n",
    "    for line in primer3_out.split('\\n'):\n",
    "        ll = line.strip().split('=')\n",
    "        if len(ll) != 2:\n",
    "            continue\n",
    "        if primer_info_start:\n",
    "            primer_dict[ll[0]] = ll[1]\n",
    "        else:\n",
    "            if ll[0].startswith('PRIMER_PAIR_0'):\n",
    "                primer_info_start = True\n",
    "                primer_dict[ll[0]] = ll[1]\n",
    "            else:\n",
    "                stat_dict[ll[0]] = ll[1]\n",
    "\n",
    "    # get primer stat df\n",
    "    primer_stat_records = []\n",
    "    primer_stat_dict = {}\n",
    "    for condition in stat_dict['PRIMER_LEFT_EXPLAIN'].split(','):\n",
    "        *condition_name, value = condition.strip().split(' ')\n",
    "        condition_name = '_'.join(condition_name)\n",
    "        primer_stat_dict[condition_name] = value\n",
    "    primer_stat_dict['primer_type'] = 'left'\n",
    "    primer_stat_records.append(pd.Series(primer_stat_dict))\n",
    "    primer_stat_dict = {}\n",
    "    for condition in stat_dict['PRIMER_RIGHT_EXPLAIN'].split(','):\n",
    "        *condition_name, value = condition.strip().split(' ')\n",
    "        condition_name = '_'.join(condition_name)\n",
    "        primer_stat_dict[condition_name] = value\n",
    "    primer_stat_dict['primer_type'] = 'right'\n",
    "    primer_stat_records.append(pd.Series(primer_stat_dict))\n",
    "    primer_stat_df = pd.DataFrame(primer_stat_records)\n",
    "    primer_stat_df['PRIMER_NAME'] = stat_dict['SEQUENCE_NAME']\n",
    "\n",
    "    # get primer df\n",
    "    primer_record_dict = collections.defaultdict(dict)\n",
    "    for k, v in primer_dict.items():\n",
    "        kl = k.split('_')\n",
    "        if len(kl) < 4:\n",
    "            continue\n",
    "        primer_id = f\"{stat_dict['SEQUENCE_NAME']}_{kl[2]}\"\n",
    "        item_name = '_'.join(kl[:2] + kl[3:])\n",
    "        primer_record_dict[primer_id][item_name] = v\n",
    "    primer_df = pd.DataFrame(primer_record_dict).T\n",
    "    primer_df['PRIMER_NAME'] = stat_dict['SEQUENCE_NAME']\n",
    "    return primer_stat_df, primer_df\n",
    "\n",
    "\n",
    "def _judge_potential_products(total_primer_df, products_cutoff=1, ratio=2):\n",
    "    valid_dict = {}\n",
    "    for primer, row in total_primer_df.iterrows():\n",
    "        potential_products = map(int, row['POTENTIAL_PRODUCT_LENGTHS'].split('|'))\n",
    "        max_length = min(int(row['PRIMER_PAIR_PRODUCT_SIZE']) * ratio, 20000)\n",
    "        valid_count = sum([i < max_length for i in potential_products])\n",
    "        valid_dict[primer] = valid_count <= products_cutoff\n",
    "    products_judge = pd.Series(valid_dict)\n",
    "    return products_judge\n",
    "\n",
    "\n",
    "def _dump_primer_fasta(total_primer_df, out_path):\n",
    "    primer_fasta = ''\n",
    "    for primer_id, sequence in total_primer_df['PRIMER_RIGHT_SEQUENCE'].iteritems():\n",
    "        fasta_record = f'>{primer_id}_r\\n{sequence}\\n'\n",
    "        primer_fasta += fasta_record\n",
    "    for primer_id, sequence in total_primer_df['PRIMER_LEFT_SEQUENCE'].iteritems():\n",
    "        fasta_record = f'>{primer_id}_l\\n{sequence}\\n'\n",
    "        primer_fasta += fasta_record\n",
    "    with open(out_path, 'w') as f:\n",
    "        f.write(primer_fasta)\n",
    "\n",
    "\n",
    "def blast_primer(primer_fasta_path,\n",
    "                 db_path,\n",
    "                 evalue_cutoff=1000,\n",
    "                 min_total_mismatch_portion=0.2,\n",
    "                 min_total_mismatch=6,\n",
    "                 min_prime_3_mismatch=2,\n",
    "                 prime_3_length=5,\n",
    "                 alt_pos_cutoff=2000,\n",
    "                 max_product_size=5000,\n",
    "                 word_size=7):\n",
    "    \"\"\"\n",
    "    Take a fasta file as input, query genome db and count qualified hits\n",
    "    :param max_product_size:\n",
    "    :param word_size:\n",
    "    :param alt_pos_cutoff:\n",
    "    :param primer_fasta_path:\n",
    "    :param db_path:\n",
    "    :param evalue_cutoff:\n",
    "    :param min_total_mismatch_portion:\n",
    "    :param min_total_mismatch:\n",
    "    :param min_prime_3_mismatch:\n",
    "    :param prime_3_length:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    # run blastn for all primers\n",
    "    temp_dir = pathlib.Path(primer_fasta_path).parent\n",
    "\n",
    "    blast_cline = NcbiblastnCommandline(query=str(primer_fasta_path),\n",
    "                                        db=db_path,\n",
    "                                        evalue=evalue_cutoff,\n",
    "                                        outfmt=5,\n",
    "                                        word_size=word_size,\n",
    "                                        out=str(temp_dir / \"primer_blast_result.xml\"),\n",
    "                                        task='blastn')\n",
    "    blast_cline()\n",
    "\n",
    "    # parse blast result\n",
    "    blast_results = SearchIO.parse(temp_dir / \"primer_blast_result.xml\", \"blast-xml\")\n",
    "    primer_hit_dict = {}\n",
    "    for blast_result in blast_results:\n",
    "        primer_length = blast_result.seq_len\n",
    "        primer_total_mismatch = max(min_total_mismatch, min_total_mismatch_portion * primer_length)\n",
    "        alternate_hsps = []\n",
    "        for hit in blast_result:\n",
    "            for hsp in hit.hsps:\n",
    "                prime_5_unmatch = [' ' for _ in range(hsp.query_range[0])]\n",
    "                prime_3_unmatch = [' ' for _ in range(primer_length - hsp.query_range[1])]\n",
    "                align_anno = prime_5_unmatch + list(hsp.aln_annotation['similarity']) + prime_3_unmatch\n",
    "                align_anno = ''.join(align_anno)\n",
    "\n",
    "                total_mismatch = primer_length - align_anno.count('|')\n",
    "                if total_mismatch > primer_total_mismatch:\n",
    "                    continue\n",
    "\n",
    "                prime_3_mismatch = prime_3_length - align_anno[-prime_3_length:].count('|')\n",
    "                if prime_3_mismatch > min_prime_3_mismatch:\n",
    "                    continue\n",
    "                alternate_hsps.append(hsp)\n",
    "        *primer_name, direction = blast_result.id.split('_')\n",
    "        primer_name = '_'.join(primer_name)\n",
    "        append_pos = 0 if direction == 'l' else 1\n",
    "        if primer_name not in primer_hit_dict:\n",
    "            primer_hit_dict[primer_name] = [[], []]\n",
    "        primer_hit_dict[primer_name][append_pos] += alternate_hsps\n",
    "    primer_hit_records = {}\n",
    "    for primer, (left_hits, right_hits) in primer_hit_dict.items():\n",
    "        if (len(left_hits) > alt_pos_cutoff) or (len(right_hits) > alt_pos_cutoff):\n",
    "            continue\n",
    "        else:\n",
    "            valid_product_lengths = []\n",
    "            positive_strand_hit = [hit for hit in left_hits if hit.hit_strand == 1] + \\\n",
    "                                  [hit for hit in right_hits if hit.hit_strand == 1]\n",
    "            negative_strand_hit = [hit for hit in left_hits if hit.hit_strand == -1] + \\\n",
    "                                  [hit for hit in right_hits if hit.hit_strand == -1]\n",
    "\n",
    "            for positive_hit in positive_strand_hit:\n",
    "                for negative_hit in negative_strand_hit:\n",
    "                    # hit not in same chrom\n",
    "                    if positive_hit.hit_id != negative_hit.hit_id:\n",
    "                        continue\n",
    "                    else:\n",
    "                        product_size = abs(positive_hit.hit_range[0] - negative_hit.hit_range[1])\n",
    "                        # left right too far away\n",
    "                        if product_size > max_product_size:\n",
    "                            continue\n",
    "\n",
    "                    valid_product_lengths.append(str(product_size))\n",
    "            primer_hit_records[primer] = {\n",
    "                'LEFT_GENOME_HITS': len(left_hits),\n",
    "                'RIGHT_GENOME_HITS': len(right_hits),\n",
    "                'POTENTIAL_PRODUCTS': len(valid_product_lengths),\n",
    "                'POTENTIAL_PRODUCT_LENGTHS': '|'.join(valid_product_lengths),\n",
    "            }\n",
    "    primer_hit_df = pd.DataFrame(primer_hit_records).T\n",
    "    return primer_hit_df\n",
    "\n",
    "\n",
    "def primer_blast(bed_path, target_fasta_path, primer3_setting_path, blast_db_path,\n",
    "                 left_expand=None, right_expand=None, both_expand=100,\n",
    "                 max_length=99999, drop_too_long=False, cpu=10, blast_kws=None,\n",
    "                 **config_kws):\n",
    "    out_dir = pathlib.Path(bed_path).parent\n",
    "\n",
    "    # parse config\n",
    "    setting_dict = {}\n",
    "    with open(primer3_setting_path) as f:\n",
    "        for line in f:\n",
    "            ll = line.strip().split('=')\n",
    "            if (len(ll) != 2) or (ll[0] == 'P3_FILE_TYPE'):\n",
    "                continue\n",
    "            setting_dict[ll[0]] = ll[1]\n",
    "    for k, v in config_kws.items():\n",
    "        setting_dict[k.upper()] = v\n",
    "\n",
    "    # all the number within primer3 config are not carefully checked, remain this to primer3.\n",
    "    if left_expand is None:\n",
    "        left_expand = both_expand\n",
    "    if right_expand is None:\n",
    "        right_expand = both_expand\n",
    "    if left_expand is None or right_expand is None:\n",
    "        raise ValueError('Specify left_expand & right_expand, or both_expand')\n",
    "\n",
    "    bed_df = _read_bed(bed_path)\n",
    "    if not pathlib.Path(target_fasta_path + '.fai').exists():\n",
    "        raise FileNotFoundError(\n",
    "            f'{target_fasta_path} do not have .fai index, use samtools faidx to index the file first')\n",
    "    fai_df = _read_fasta_fai(target_fasta_path + '.fai')\n",
    "\n",
    "    primer_records = []\n",
    "    for primer_name, (seq_name, start, end) in bed_df.iterrows():\n",
    "        region_length = start - end\n",
    "        total_length = left_expand + region_length + right_expand\n",
    "        if region_length > max_length:\n",
    "            print(f'{primer_name} is dropped due to exceed {max_length} bp')\n",
    "            continue\n",
    "        elif total_length > max_length:\n",
    "            if drop_too_long:\n",
    "                print(f'{primer_name} is dropped due to exceed {max_length} bp (including expansion)')\n",
    "                continue\n",
    "            print(f'{primer_name} expansion is shorten due to exceed {max_length} bp')\n",
    "            extra_length = max_length - region_length\n",
    "            left_portion = left_expand / (left_expand + right_expand)\n",
    "            left_expand = int(extra_length * left_portion)\n",
    "            right_expand = extra_length - left_expand\n",
    "\n",
    "        primer_series = _query_genome(fasta_path=target_fasta_path,\n",
    "                                      fai_df=fai_df,\n",
    "                                      seq_name=seq_name,\n",
    "                                      region_start=start,\n",
    "                                      region_end=end,\n",
    "                                      left_expand=left_expand,\n",
    "                                      right_expand=right_expand,\n",
    "                                      primer_name=primer_name)\n",
    "        primer_records.append(primer_series)\n",
    "    primer_template_df = pd.DataFrame(primer_records).set_index('SEQUENCE_NAME')\n",
    "    primer_template_df, total_primer_stat_df, total_primer_df = _run_primer3(primer_template_df,\n",
    "                                                                             setting_dict, cpu=cpu)\n",
    "\n",
    "    _dump_primer_fasta(total_primer_df, out_dir / 'primer.fa')\n",
    "\n",
    "    if blast_kws is None:\n",
    "        blast_kws = {}\n",
    "    primer_hit_df = blast_primer(primer_fasta_path=out_dir / 'primer.fa',\n",
    "                                 db_path=blast_db_path,\n",
    "                                 **blast_kws)\n",
    "    total_primer_df = pd.concat([total_primer_df, primer_hit_df], axis=1, sort=True)\n",
    "    total_primer_df = total_primer_df.reindex(primer_hit_df.index)\n",
    "\n",
    "    potential_product_judge = _judge_potential_products(total_primer_df, products_cutoff=1, ratio=3)\n",
    "    filtered_primer_df = total_primer_df.loc[potential_product_judge]\n",
    "    selected_primer = filtered_primer_df[['PRIMER_RIGHT_PENALTY', 'PRIMER_LEFT_PENALTY']] \\\n",
    "        .astype(float) \\\n",
    "        .sum(axis=1) \\\n",
    "        .groupby(filtered_primer_df.index.map(lambda i: i.split('_')[0])) \\\n",
    "        .apply(lambda sub_df: sub_df.sort_values().index[0]).tolist()\n",
    "\n",
    "    total_primer_df['Selected'] = total_primer_df.index.map(\n",
    "        lambda i: True if i in set(selected_primer) else False)\n",
    "\n",
    "    primer_template_df.to_csv(out_dir / 'primer_template.tsv.gz', sep='\\t', compression='gzip')\n",
    "    total_primer_stat_df.to_csv(out_dir / 'primer3_stat.tsv.gz', sep='\\t', compression='gzip')\n",
    "    total_primer_df.to_csv(out_dir / 'primer.tsv.gz', sep='\\t', compression='gzip')\n",
    "    subprocess.run(['rm', str(out_dir / 'primer_blast_result.xml')])\n",
    "    return primer_template_df, total_primer_stat_df, total_primer_df, selected_primer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-30T07:35:05.146737Z",
     "start_time": "2018-11-30T07:32:46.999851Z"
    }
   },
   "outputs": [],
   "source": [
    "primer_template_df, total_primer_stat_df, total_primer_df, selected_primer= primer_blast(\n",
    "    bed_path='./test_region.mm10.bed', \n",
    "    target_fasta_path='/home/hanliu/ref/mouse/genome/fasta/with_chrl/mm10_with_chrl.fa', \n",
    "    primer3_setting_path='./primer3_settings.txt',\n",
    "    blast_db_path='/home/hanliu/ref/mouse/genome/fasta/raw/mm10.fa',\n",
    "    left_expand=None, \n",
    "    right_expand=None, \n",
    "    both_expand=150,\n",
    "    max_length=99999, \n",
    "    drop_too_long=False, \n",
    "    cpu=10,\n",
    "    PRIMER_MISPRIMING_LIBRARY='/home/hanliu/pkg/primer_design/repeat/rodent_repeat_library.fa',\n",
    "    PRIMER_PRODUCT_SIZE_RANGE='40-20000',\n",
    "    PRIMER_NUM_RETURN='100',\n",
    "    PRIMER_THERMODYNAMIC_PARAMETERS_PATH='/home/hanliu/anaconda3/pkgs/primer3-2.4.1a-pl526hfc679d8_0/share/primer3/primer3_config/',\n",
    "    PRIMER_MASK_KMERLIST_PATH='/home/hanliu/ref/mouse/primer_3_kmer/',\n",
    "    PRIMER_MASK_KMERLIST_PREFIX='mus_musculus',\n",
    "    PRIMER_LOWERCASE_MASKING='0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-11-30T09:05:43.564325Z",
     "start_time": "2018-11-30T09:05:43.522092Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PRIMER_LEFT_END_STABILITY</th>\n",
       "      <th>PRIMER_LEFT_GC_PERCENT</th>\n",
       "      <th>PRIMER_LEFT_HAIRPIN_TH</th>\n",
       "      <th>PRIMER_LEFT_LIBRARY_MISPRIMING</th>\n",
       "      <th>PRIMER_LEFT_PENALTY</th>\n",
       "      <th>PRIMER_LEFT_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_LEFT_SELF_END_TH</th>\n",
       "      <th>PRIMER_LEFT_SEQUENCE</th>\n",
       "      <th>PRIMER_LEFT_TEMPLATE_MISPRIMING</th>\n",
       "      <th>PRIMER_LEFT_TM</th>\n",
       "      <th>...</th>\n",
       "      <th>PRIMER_RIGHT_SELF_ANY_TH</th>\n",
       "      <th>PRIMER_RIGHT_SELF_END_TH</th>\n",
       "      <th>PRIMER_RIGHT_SEQUENCE</th>\n",
       "      <th>PRIMER_RIGHT_TEMPLATE_MISPRIMING</th>\n",
       "      <th>PRIMER_RIGHT_TM</th>\n",
       "      <th>LEFT_GENOME_HITS</th>\n",
       "      <th>POTENTIAL_PRODUCTS</th>\n",
       "      <th>POTENTIAL_PRODUCT_LENGTHS</th>\n",
       "      <th>RIGHT_GENOME_HITS</th>\n",
       "      <th>Selected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1A1_0</th>\n",
       "      <td>3.0100</td>\n",
       "      <td>60.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MCMREP Murine middle repetitive DNA.</td>\n",
       "      <td>0.036858</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTGACCTCCCTGAGCAGTTC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CACCCACTACACCCCTCAAC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.964</td>\n",
       "      <td>428</td>\n",
       "      <td>3</td>\n",
       "      <td>646|4624|2654</td>\n",
       "      <td>344</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_1</th>\n",
       "      <td>4.2000</td>\n",
       "      <td>55.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.00, MMLTR1 Mouse long terminal repeat.</td>\n",
       "      <td>0.474660</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AAGCTTCCAGTGACCTCCCT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.475</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCTCCTGTTTCCTCACCCAC</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>59.963</td>\n",
       "      <td>623</td>\n",
       "      <td>9</td>\n",
       "      <td>668|4341|2279|4246|2675|4363|3210|2022|4106</td>\n",
       "      <td>1477</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_2</th>\n",
       "      <td>4.2600</td>\n",
       "      <td>60.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, reverse ORR1A Rodent ORR1a MaLR-LTR - a...</td>\n",
       "      <td>1.194781</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TTCCAGTGACCTCCCTGAGC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>61.195</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TTGGAGAGGCTAGGAGGGAG</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.737</td>\n",
       "      <td>617</td>\n",
       "      <td>6</td>\n",
       "      <td>749|4266|1027|2513|2491|1249</td>\n",
       "      <td>1401</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_4</th>\n",
       "      <td>4.4100</td>\n",
       "      <td>47.619</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, MML1 Mouse L1 repetitive sequence.</td>\n",
       "      <td>1.748493</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTGTCCCAACTGAAAACAGCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.252</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TTCCTCACCCACTACACCCC</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>60.839</td>\n",
       "      <td>616</td>\n",
       "      <td>3</td>\n",
       "      <td>715|4323|329</td>\n",
       "      <td>606</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_5</th>\n",
       "      <td>4.0200</td>\n",
       "      <td>54.545</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, MMIAPA Repeat associated with intracist...</td>\n",
       "      <td>2.158798</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.58</td>\n",
       "      <td>GGGGTATAAGCTTCCAGTGACC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>60.159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCACTACACCCCTCAACTGT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>58.943</td>\n",
       "      <td>149</td>\n",
       "      <td>1</td>\n",
       "      <td>659</td>\n",
       "      <td>300</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_6</th>\n",
       "      <td>2.5700</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MML1 Mouse L1 repetitive sequence.</td>\n",
       "      <td>2.894126</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CAGGAACACACAGAGACCTAGT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.106</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCTGTTTCCTCACCCACTACA</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>59.303</td>\n",
       "      <td>955</td>\n",
       "      <td>6</td>\n",
       "      <td>604|4059|2650|3563|78|743</td>\n",
       "      <td>583</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_7</th>\n",
       "      <td>2.9200</td>\n",
       "      <td>45.455</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, MMZP3A Repetitive segment of ZP3 gene.</td>\n",
       "      <td>3.095332</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGCAGTTCTCAGTAGCAGATCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>58.905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCAGAGAAAGTGACCCCAGAAT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.291</td>\n",
       "      <td>443</td>\n",
       "      <td>3</td>\n",
       "      <td>753|920|101</td>\n",
       "      <td>685</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_8</th>\n",
       "      <td>3.8200</td>\n",
       "      <td>45.455</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, MMSQR2 Mouse simple repetitive DNA (sqr...</td>\n",
       "      <td>3.206150</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACATCAAGACACAGGAACACAC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>58.794</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CAGAATTGGAGAGGCTAGGAGG</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.632</td>\n",
       "      <td>1209</td>\n",
       "      <td>1</td>\n",
       "      <td>708</td>\n",
       "      <td>363</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A1_9</th>\n",
       "      <td>4.2600</td>\n",
       "      <td>45.455</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MMLTR5 Mouse endogenous retrovirus (MuR...</td>\n",
       "      <td>3.762779</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCATGGTCATTGTAAGTTCAGC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>58.237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTACAAGCAGAGAAAGTGACCC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>58.670</td>\n",
       "      <td>374</td>\n",
       "      <td>4</td>\n",
       "      <td>817|3601|749|1815</td>\n",
       "      <td>546</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A3_0</th>\n",
       "      <td>4.8500</td>\n",
       "      <td>60.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, reverse PMR89 Mouse PMR89 middle repeti...</td>\n",
       "      <td>0.534873</td>\n",
       "      <td>7.85</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGACTGTCTTCTCTCCAGC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.465</td>\n",
       "      <td>...</td>\n",
       "      <td>18.61</td>\n",
       "      <td>2.22</td>\n",
       "      <td>CACATCATAGGCAGTGCAAAGT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.249</td>\n",
       "      <td>1359</td>\n",
       "      <td>6</td>\n",
       "      <td>1416|711|276|594|4270|4225</td>\n",
       "      <td>275</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A3_1</th>\n",
       "      <td>4.4000</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.00, MMCHYREP Mus musculus (clone C573) chrom...</td>\n",
       "      <td>2.159627</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CGGATTGGTTGTTGACTAGTGC</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>59.840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCTAGTCACATCATAGGCAGTG</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>58.612</td>\n",
       "      <td>132</td>\n",
       "      <td>3</td>\n",
       "      <td>1474|4023|4804</td>\n",
       "      <td>326</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A5_0</th>\n",
       "      <td>2.9700</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, MAIAPLTR Syrian hamster intracisternal ...</td>\n",
       "      <td>3.252351</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACCCCATCCCTTAGAAGATAGC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>58.748</td>\n",
       "      <td>...</td>\n",
       "      <td>20.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACCTGCATGCTATAAGAGGGC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.927</td>\n",
       "      <td>295</td>\n",
       "      <td>22</td>\n",
       "      <td>1358|1355|1355|1358|1355|4206|2825|2779|2113|2...</td>\n",
       "      <td>1801</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A8_0</th>\n",
       "      <td>5.9200</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12.00, MMIAPLTR1 Mouse intracisternal IAP 3' LTR.</td>\n",
       "      <td>0.657840</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AAAGAACACACTGTCAGCGC</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>59.342</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TGGGACACTGGGACACTACT</td>\n",
       "      <td>12.0000</td>\n",
       "      <td>59.810</td>\n",
       "      <td>340</td>\n",
       "      <td>1</td>\n",
       "      <td>1378</td>\n",
       "      <td>377</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A8_1</th>\n",
       "      <td>3.5800</td>\n",
       "      <td>47.619</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MML1 Mouse L1 repetitive sequence.</td>\n",
       "      <td>1.401552</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCTGCAAAGAACACACTGTCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGTCAGTGCCTCTTCTCTGG</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.751</td>\n",
       "      <td>842</td>\n",
       "      <td>7</td>\n",
       "      <td>1420|1410|2526|1786|1948|2193|3195</td>\n",
       "      <td>1164</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A8_2</th>\n",
       "      <td>5.1900</td>\n",
       "      <td>55.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, reverse MAIAPLTR Syrian hamster intraci...</td>\n",
       "      <td>1.475403</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGAAAAGCTAAAGGAGGGC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>58.525</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAAGAGCACCCACTGAAGCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>60.250</td>\n",
       "      <td>776</td>\n",
       "      <td>5</td>\n",
       "      <td>1908|1399|1670|3669|3672</td>\n",
       "      <td>524</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A8_3</th>\n",
       "      <td>3.5500</td>\n",
       "      <td>45.455</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MAIAPLTR Syrian hamster intracisternal ...</td>\n",
       "      <td>2.033801</td>\n",
       "      <td>6.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTGATGCTGCAAAGAACACACT</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.966</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ATCGACCTCCCCTGCCTTT</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>60.617</td>\n",
       "      <td>1311</td>\n",
       "      <td>4</td>\n",
       "      <td>1476|4748|2118|4294</td>\n",
       "      <td>1047</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A8_4</th>\n",
       "      <td>3.8500</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MMREPECORI Mouse repetitive sequence (c...</td>\n",
       "      <td>2.424956</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGTTCATGATGCTTTGGACTCC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.575</td>\n",
       "      <td>...</td>\n",
       "      <td>20.43</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCTTTGCAGCTGATTCGAAGA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>58.921</td>\n",
       "      <td>283</td>\n",
       "      <td>1</td>\n",
       "      <td>1358</td>\n",
       "      <td>159</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A8_5</th>\n",
       "      <td>3.3600</td>\n",
       "      <td>45.455</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MML1 Mouse L1 repetitive sequence.</td>\n",
       "      <td>2.986728</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGTGAGGGGAAAAGCTAAAGGA</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.013</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TCGAAGAGCACCCACTGAA</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>58.563</td>\n",
       "      <td>953</td>\n",
       "      <td>4</td>\n",
       "      <td>1407|2778|4906|3731</td>\n",
       "      <td>647</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A9_0</th>\n",
       "      <td>3.5100</td>\n",
       "      <td>60.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MMLTR2 Mouse endogenous retroviral DNA,...</td>\n",
       "      <td>0.177604</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCCTCCTGACTCATGCACTC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.822</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GACCCCTCCTTGCAATCTGG</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>60.395</td>\n",
       "      <td>501</td>\n",
       "      <td>3</td>\n",
       "      <td>1304|470|4100</td>\n",
       "      <td>335</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A9_1</th>\n",
       "      <td>2.9000</td>\n",
       "      <td>57.895</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MMLTR1 Mouse long terminal repeat.</td>\n",
       "      <td>1.610673</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACACGCCCTCCTGACTCAT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.611</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCTTGCAATCTGGGTGGG</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.682</td>\n",
       "      <td>597</td>\n",
       "      <td>2</td>\n",
       "      <td>1304|425</td>\n",
       "      <td>273</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A9_2</th>\n",
       "      <td>4.1600</td>\n",
       "      <td>45.455</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12.00, MMYREPA11 Mus musculus (clone A11) chro...</td>\n",
       "      <td>3.599348</td>\n",
       "      <td>6.56</td>\n",
       "      <td>6.56</td>\n",
       "      <td>AACCCCTTTTCTAGAGAGTGGT</td>\n",
       "      <td>10.0000</td>\n",
       "      <td>58.401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACCATCTGTTCACCCTCCTAGT</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>60.226</td>\n",
       "      <td>402</td>\n",
       "      <td>2</td>\n",
       "      <td>1408|2364</td>\n",
       "      <td>307</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A11_0</th>\n",
       "      <td>3.9500</td>\n",
       "      <td>55.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, reverse MML1 Mouse L1 repetitive sequence.</td>\n",
       "      <td>0.713579</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGGAGAGACTACTGCCCCTT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.286</td>\n",
       "      <td>...</td>\n",
       "      <td>5.17</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TCTGTGTCTGACCCAAAGCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.163</td>\n",
       "      <td>229</td>\n",
       "      <td>2</td>\n",
       "      <td>1175|869</td>\n",
       "      <td>615</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B1_0</th>\n",
       "      <td>3.1100</td>\n",
       "      <td>55.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, reverse RNL1R2 Rat L1R2 sequence DNA.</td>\n",
       "      <td>0.111253</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGTGGGAAGTGGCTCAAAGG</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>59.889</td>\n",
       "      <td>...</td>\n",
       "      <td>16.43</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TCTTCTGAAGTAGGCCACGC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.753</td>\n",
       "      <td>652</td>\n",
       "      <td>3</td>\n",
       "      <td>1157|750|4123</td>\n",
       "      <td>152</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B1_1</th>\n",
       "      <td>4.2600</td>\n",
       "      <td>55.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12.00, reverse PLMYS1 Mouse mys-1 transposon.</td>\n",
       "      <td>0.178921</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGAGAGTGGGAAGTGGCTCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>60.179</td>\n",
       "      <td>...</td>\n",
       "      <td>15.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGCCACGCGATCATCACTAT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.320</td>\n",
       "      <td>729</td>\n",
       "      <td>1</td>\n",
       "      <td>1149</td>\n",
       "      <td>32</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B1_2</th>\n",
       "      <td>5.0100</td>\n",
       "      <td>57.143</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, reverse PLMYS1 Mouse mys-1 transposon.</td>\n",
       "      <td>1.205893</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTGTAGAGAGTGGGAAGTGGC</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>59.794</td>\n",
       "      <td>...</td>\n",
       "      <td>15.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AAGTAGGCCACGCGATCATC</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>60.249</td>\n",
       "      <td>806</td>\n",
       "      <td>1</td>\n",
       "      <td>1158</td>\n",
       "      <td>28</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B1_3</th>\n",
       "      <td>2.3200</td>\n",
       "      <td>52.381</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.00, reverse MGR MGR repeat.</td>\n",
       "      <td>1.331756</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCCAGTGCGATGTCCTAAATG</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>59.668</td>\n",
       "      <td>...</td>\n",
       "      <td>15.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTGAAGTAGGCCACGCGATC</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>60.872</td>\n",
       "      <td>133</td>\n",
       "      <td>2</td>\n",
       "      <td>1125|2597</td>\n",
       "      <td>27</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B1_4</th>\n",
       "      <td>3.5600</td>\n",
       "      <td>57.895</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.00, MML1 Mouse L1 repetitive sequence.</td>\n",
       "      <td>1.514553</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCTCAAAGGGCAGGAATGC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.485</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGTGGGCTTCTAAATGTCCTGA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.087</td>\n",
       "      <td>800</td>\n",
       "      <td>2</td>\n",
       "      <td>1190|756</td>\n",
       "      <td>341</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B3_0</th>\n",
       "      <td>3.5500</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>12.00, MMREPECORI Mouse repetitive sequence (c...</td>\n",
       "      <td>0.258774</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AGGGTTTGGCTTGTCAGTGT</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.741</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACTGCGTGGGGTTTTGGTAA</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>60.107</td>\n",
       "      <td>555</td>\n",
       "      <td>1</td>\n",
       "      <td>1159</td>\n",
       "      <td>201</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B3_1</th>\n",
       "      <td>5.1900</td>\n",
       "      <td>60.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, PR Novel repeat element (PR).</td>\n",
       "      <td>0.321122</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTTGTCAGTGTCTCCAGCCC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>60.321</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CGTGGGGTTTTGGTAATTTGCT</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.963</td>\n",
       "      <td>916</td>\n",
       "      <td>3</td>\n",
       "      <td>1146|1522|3715</td>\n",
       "      <td>439</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B3_2</th>\n",
       "      <td>3.8600</td>\n",
       "      <td>55.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>11.00, reverse RAL Rat RAL repetitive element.</td>\n",
       "      <td>0.394192</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TGGCTTGTCAGTGTCTCCAG</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.606</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTGCGTGGGGTTTTGGTAATTT</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.963</td>\n",
       "      <td>733</td>\n",
       "      <td>1</td>\n",
       "      <td>1152</td>\n",
       "      <td>209</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B3_3</th>\n",
       "      <td>4.3000</td>\n",
       "      <td>60.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, MTC Rodent MTc MaLR-LTR - a consensus</td>\n",
       "      <td>0.831307</td>\n",
       "      <td>2.58</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCCCAGCAAGTAGTACTCCC</td>\n",
       "      <td>7.0000</td>\n",
       "      <td>59.169</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AATGTACTCCAAACACTGCGT</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>58.154</td>\n",
       "      <td>236</td>\n",
       "      <td>1</td>\n",
       "      <td>1147</td>\n",
       "      <td>183</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B3_4</th>\n",
       "      <td>4.9200</td>\n",
       "      <td>52.381</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, RAL Rat RAL repetitive element.</td>\n",
       "      <td>1.278127</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTAACCTCTGCTTACCTGCCA</td>\n",
       "      <td>9.0000</td>\n",
       "      <td>59.722</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TGTCCCAACACAAAGTCACAA</td>\n",
       "      <td>10.0000</td>\n",
       "      <td>58.203</td>\n",
       "      <td>539</td>\n",
       "      <td>6</td>\n",
       "      <td>1094|1491|579|2597|1824|1512</td>\n",
       "      <td>824</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B4_0</th>\n",
       "      <td>4.2600</td>\n",
       "      <td>52.381</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10.00, reverse RNLPKG Repetitive segment of L-...</td>\n",
       "      <td>1.208052</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCATACTTCAGACCTGGCTCA</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TCCCTTTGCTAACTTTGTGGC</td>\n",
       "      <td>8.0000</td>\n",
       "      <td>59.039</td>\n",
       "      <td>606</td>\n",
       "      <td>5</td>\n",
       "      <td>1112|4966|4389|3703|3403</td>\n",
       "      <td>626</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1B4_1</th>\n",
       "      <td>4.9500</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9.00, MMLTR5 Mouse endogenous retrovirus (MuRV...</td>\n",
       "      <td>1.555967</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TTGCATTTTCTGGCTACCCC</td>\n",
       "      <td>6.0000</td>\n",
       "      <td>58.444</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACCAGACTTTCCCTTTGCTAAC</td>\n",
       "      <td>10.0000</td>\n",
       "      <td>58.513</td>\n",
       "      <td>369</td>\n",
       "      <td>2</td>\n",
       "      <td>1178|3693</td>\n",
       "      <td>371</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34 rows  32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PRIMER_LEFT_END_STABILITY PRIMER_LEFT_GC_PERCENT  \\\n",
       "1A1_0                     3.0100                 60.000   \n",
       "1A1_1                     4.2000                 55.000   \n",
       "1A1_2                     4.2600                 60.000   \n",
       "1A1_4                     4.4100                 47.619   \n",
       "1A1_5                     4.0200                 54.545   \n",
       "1A1_6                     2.5700                 50.000   \n",
       "1A1_7                     2.9200                 45.455   \n",
       "1A1_8                     3.8200                 45.455   \n",
       "1A1_9                     4.2600                 45.455   \n",
       "1A3_0                     4.8500                 60.000   \n",
       "1A3_1                     4.4000                 50.000   \n",
       "1A5_0                     2.9700                 50.000   \n",
       "1A8_0                     5.9200                 50.000   \n",
       "1A8_1                     3.5800                 47.619   \n",
       "1A8_2                     5.1900                 55.000   \n",
       "1A8_3                     3.5500                 45.455   \n",
       "1A8_4                     3.8500                 50.000   \n",
       "1A8_5                     3.3600                 45.455   \n",
       "1A9_0                     3.5100                 60.000   \n",
       "1A9_1                     2.9000                 57.895   \n",
       "1A9_2                     4.1600                 45.455   \n",
       "1A11_0                    3.9500                 55.000   \n",
       "1B1_0                     3.1100                 55.000   \n",
       "1B1_1                     4.2600                 55.000   \n",
       "1B1_2                     5.0100                 57.143   \n",
       "1B1_3                     2.3200                 52.381   \n",
       "1B1_4                     3.5600                 57.895   \n",
       "1B3_0                     3.5500                 50.000   \n",
       "1B3_1                     5.1900                 60.000   \n",
       "1B3_2                     3.8600                 55.000   \n",
       "1B3_3                     4.3000                 60.000   \n",
       "1B3_4                     4.9200                 52.381   \n",
       "1B4_0                     4.2600                 52.381   \n",
       "1B4_1                     4.9500                 50.000   \n",
       "\n",
       "       PRIMER_LEFT_HAIRPIN_TH  \\\n",
       "1A1_0                    0.00   \n",
       "1A1_1                    0.00   \n",
       "1A1_2                    0.00   \n",
       "1A1_4                    0.00   \n",
       "1A1_5                    0.00   \n",
       "1A1_6                    0.00   \n",
       "1A1_7                    0.00   \n",
       "1A1_8                    0.00   \n",
       "1A1_9                    0.00   \n",
       "1A3_0                    0.00   \n",
       "1A3_1                    0.00   \n",
       "1A5_0                    0.00   \n",
       "1A8_0                    0.00   \n",
       "1A8_1                    0.00   \n",
       "1A8_2                    0.00   \n",
       "1A8_3                    0.00   \n",
       "1A8_4                    0.00   \n",
       "1A8_5                    0.00   \n",
       "1A9_0                    0.00   \n",
       "1A9_1                    0.00   \n",
       "1A9_2                    0.00   \n",
       "1A11_0                   0.00   \n",
       "1B1_0                    0.00   \n",
       "1B1_1                    0.00   \n",
       "1B1_2                    0.00   \n",
       "1B1_3                    0.00   \n",
       "1B1_4                    0.00   \n",
       "1B3_0                    0.00   \n",
       "1B3_1                    0.00   \n",
       "1B3_2                    0.00   \n",
       "1B3_3                    0.00   \n",
       "1B3_4                    0.00   \n",
       "1B4_0                    0.00   \n",
       "1B4_1                    0.00   \n",
       "\n",
       "                           PRIMER_LEFT_LIBRARY_MISPRIMING PRIMER_LEFT_PENALTY  \\\n",
       "1A1_0         10.00, MCMREP Murine middle repetitive DNA.            0.036858   \n",
       "1A1_1            9.00, MMLTR1 Mouse long terminal repeat.            0.474660   \n",
       "1A1_2   10.00, reverse ORR1A Rodent ORR1a MaLR-LTR - a...            1.194781   \n",
       "1A1_4           11.00, MML1 Mouse L1 repetitive sequence.            1.748493   \n",
       "1A1_5   11.00, MMIAPA Repeat associated with intracist...            2.158798   \n",
       "1A1_6           10.00, MML1 Mouse L1 repetitive sequence.            2.894126   \n",
       "1A1_7       11.00, MMZP3A Repetitive segment of ZP3 gene.            3.095332   \n",
       "1A1_8   11.00, MMSQR2 Mouse simple repetitive DNA (sqr...            3.206150   \n",
       "1A1_9   10.00, MMLTR5 Mouse endogenous retrovirus (MuR...            3.762779   \n",
       "1A3_0   11.00, reverse PMR89 Mouse PMR89 middle repeti...            0.534873   \n",
       "1A3_1   9.00, MMCHYREP Mus musculus (clone C573) chrom...            2.159627   \n",
       "1A5_0   11.00, MAIAPLTR Syrian hamster intracisternal ...            3.252351   \n",
       "1A8_0   12.00, MMIAPLTR1 Mouse intracisternal IAP 3' LTR.            0.657840   \n",
       "1A8_1           10.00, MML1 Mouse L1 repetitive sequence.            1.401552   \n",
       "1A8_2   10.00, reverse MAIAPLTR Syrian hamster intraci...            1.475403   \n",
       "1A8_3   10.00, MAIAPLTR Syrian hamster intracisternal ...            2.033801   \n",
       "1A8_4   10.00, MMREPECORI Mouse repetitive sequence (c...            2.424956   \n",
       "1A8_5           10.00, MML1 Mouse L1 repetitive sequence.            2.986728   \n",
       "1A9_0   10.00, MMLTR2 Mouse endogenous retroviral DNA,...            0.177604   \n",
       "1A9_1           10.00, MMLTR1 Mouse long terminal repeat.            1.610673   \n",
       "1A9_2   12.00, MMYREPA11 Mus musculus (clone A11) chro...            3.599348   \n",
       "1A11_0  10.00, reverse MML1 Mouse L1 repetitive sequence.            0.713579   \n",
       "1B1_0        10.00, reverse RNL1R2 Rat L1R2 sequence DNA.            0.111253   \n",
       "1B1_1       12.00, reverse PLMYS1 Mouse mys-1 transposon.            0.178921   \n",
       "1B1_2       11.00, reverse PLMYS1 Mouse mys-1 transposon.            1.205893   \n",
       "1B1_3                       9.00, reverse MGR MGR repeat.            1.331756   \n",
       "1B1_4            9.00, MML1 Mouse L1 repetitive sequence.            1.514553   \n",
       "1B3_0   12.00, MMREPECORI Mouse repetitive sequence (c...            0.258774   \n",
       "1B3_1                11.00, PR Novel repeat element (PR).            0.321122   \n",
       "1B3_2      11.00, reverse RAL Rat RAL repetitive element.            0.394192   \n",
       "1B3_3        10.00, MTC Rodent MTc MaLR-LTR - a consensus            0.831307   \n",
       "1B3_4              10.00, RAL Rat RAL repetitive element.            1.278127   \n",
       "1B4_0   10.00, reverse RNLPKG Repetitive segment of L-...            1.208052   \n",
       "1B4_1   9.00, MMLTR5 Mouse endogenous retrovirus (MuRV...            1.555967   \n",
       "\n",
       "       PRIMER_LEFT_SELF_ANY_TH PRIMER_LEFT_SELF_END_TH  \\\n",
       "1A1_0                     0.00                    0.00   \n",
       "1A1_1                     0.00                    0.00   \n",
       "1A1_2                     0.00                    0.00   \n",
       "1A1_4                     0.00                    0.00   \n",
       "1A1_5                     0.00                    2.58   \n",
       "1A1_6                     0.00                    0.00   \n",
       "1A1_7                     0.00                    0.00   \n",
       "1A1_8                     0.00                    0.00   \n",
       "1A1_9                     0.00                    0.00   \n",
       "1A3_0                     7.85                    0.00   \n",
       "1A3_1                     0.00                    0.00   \n",
       "1A5_0                     0.00                    0.00   \n",
       "1A8_0                     0.00                    0.00   \n",
       "1A8_1                     0.00                    0.00   \n",
       "1A8_2                     0.00                    0.00   \n",
       "1A8_3                     6.36                    0.00   \n",
       "1A8_4                     0.00                    0.00   \n",
       "1A8_5                     0.00                    0.00   \n",
       "1A9_0                     0.00                    0.00   \n",
       "1A9_1                     0.00                    0.00   \n",
       "1A9_2                     6.56                    6.56   \n",
       "1A11_0                    0.00                    0.00   \n",
       "1B1_0                     0.00                    0.00   \n",
       "1B1_1                     0.00                    0.00   \n",
       "1B1_2                     0.00                    0.00   \n",
       "1B1_3                     0.00                    0.00   \n",
       "1B1_4                     0.00                    0.00   \n",
       "1B3_0                     0.00                    0.00   \n",
       "1B3_1                     0.00                    0.00   \n",
       "1B3_2                     0.00                    0.00   \n",
       "1B3_3                     2.58                    0.00   \n",
       "1B3_4                     0.00                    0.00   \n",
       "1B4_0                     0.00                    0.00   \n",
       "1B4_1                     0.00                    0.00   \n",
       "\n",
       "          PRIMER_LEFT_SEQUENCE PRIMER_LEFT_TEMPLATE_MISPRIMING PRIMER_LEFT_TM  \\\n",
       "1A1_0     GTGACCTCCCTGAGCAGTTC                          7.0000         60.037   \n",
       "1A1_1     AAGCTTCCAGTGACCTCCCT                          7.0000         60.475   \n",
       "1A1_2     TTCCAGTGACCTCCCTGAGC                          7.0000         61.195   \n",
       "1A1_4    GTGTCCCAACTGAAAACAGCA                          8.0000         59.252   \n",
       "1A1_5   GGGGTATAAGCTTCCAGTGACC                          8.0000         60.159   \n",
       "1A1_6   CAGGAACACACAGAGACCTAGT                          7.0000         59.106   \n",
       "1A1_7   AGCAGTTCTCAGTAGCAGATCA                          8.0000         58.905   \n",
       "1A1_8   ACATCAAGACACAGGAACACAC                          8.0000         58.794   \n",
       "1A1_9   GCATGGTCATTGTAAGTTCAGC                          7.0000         58.237   \n",
       "1A3_0     GGGACTGTCTTCTCTCCAGC                          8.0000         59.465   \n",
       "1A3_1   CGGATTGGTTGTTGACTAGTGC                          9.0000         59.840   \n",
       "1A5_0   ACCCCATCCCTTAGAAGATAGC                          7.0000         58.748   \n",
       "1A8_0     AAAGAACACACTGTCAGCGC                          9.0000         59.342   \n",
       "1A8_1    GCTGCAAAGAACACACTGTCA                          8.0000         59.598   \n",
       "1A8_2     GGGAAAAGCTAAAGGAGGGC                          7.0000         58.525   \n",
       "1A8_3   GTGATGCTGCAAAGAACACACT                          8.0000         59.966   \n",
       "1A8_4   GGTTCATGATGCTTTGGACTCC                          8.0000         59.575   \n",
       "1A8_5   AGTGAGGGGAAAAGCTAAAGGA                          7.0000         59.013   \n",
       "1A9_0     CCCTCCTGACTCATGCACTC                          8.0000         59.822   \n",
       "1A9_1      ACACGCCCTCCTGACTCAT                          7.0000         60.611   \n",
       "1A9_2   AACCCCTTTTCTAGAGAGTGGT                         10.0000         58.401   \n",
       "1A11_0    AGGAGAGACTACTGCCCCTT                          7.0000         59.286   \n",
       "1B1_0     AGTGGGAAGTGGCTCAAAGG                          9.0000         59.889   \n",
       "1B1_1     AGAGAGTGGGAAGTGGCTCA                          8.0000         60.179   \n",
       "1B1_2    CTGTAGAGAGTGGGAAGTGGC                          9.0000         59.794   \n",
       "1B1_3    GCCAGTGCGATGTCCTAAATG                          9.0000         59.668   \n",
       "1B1_4      GCTCAAAGGGCAGGAATGC                          7.0000         59.485   \n",
       "1B3_0     AGGGTTTGGCTTGTCAGTGT                          7.0000         59.741   \n",
       "1B3_1     CTTGTCAGTGTCTCCAGCCC                          7.0000         60.321   \n",
       "1B3_2     TGGCTTGTCAGTGTCTCCAG                          8.0000         59.606   \n",
       "1B3_3     CCCCAGCAAGTAGTACTCCC                          7.0000         59.169   \n",
       "1B3_4    GTAACCTCTGCTTACCTGCCA                          9.0000         59.722   \n",
       "1B4_0    GCATACTTCAGACCTGGCTCA                          8.0000         59.792   \n",
       "1B4_1     TTGCATTTTCTGGCTACCCC                          6.0000         58.444   \n",
       "\n",
       "         ...    PRIMER_RIGHT_SELF_ANY_TH PRIMER_RIGHT_SELF_END_TH  \\\n",
       "1A1_0    ...                        0.00                     0.00   \n",
       "1A1_1    ...                        0.00                     0.00   \n",
       "1A1_2    ...                        0.00                     0.00   \n",
       "1A1_4    ...                        0.00                     0.00   \n",
       "1A1_5    ...                        0.00                     0.00   \n",
       "1A1_6    ...                        0.00                     0.00   \n",
       "1A1_7    ...                        0.00                     0.00   \n",
       "1A1_8    ...                        0.00                     0.00   \n",
       "1A1_9    ...                        0.00                     0.00   \n",
       "1A3_0    ...                       18.61                     2.22   \n",
       "1A3_1    ...                        0.00                     0.00   \n",
       "1A5_0    ...                       20.33                     0.00   \n",
       "1A8_0    ...                        0.00                     0.00   \n",
       "1A8_1    ...                        0.00                     0.00   \n",
       "1A8_2    ...                        0.00                     0.00   \n",
       "1A8_3    ...                        0.00                     0.00   \n",
       "1A8_4    ...                       20.43                     0.00   \n",
       "1A8_5    ...                        0.00                     0.00   \n",
       "1A9_0    ...                        0.00                     0.00   \n",
       "1A9_1    ...                        0.00                     0.00   \n",
       "1A9_2    ...                        0.00                     0.00   \n",
       "1A11_0   ...                        5.17                     0.00   \n",
       "1B1_0    ...                       16.43                     0.00   \n",
       "1B1_1    ...                       15.93                     0.00   \n",
       "1B1_2    ...                       15.93                     0.00   \n",
       "1B1_3    ...                       15.93                     0.00   \n",
       "1B1_4    ...                        0.00                     0.00   \n",
       "1B3_0    ...                        0.00                     0.00   \n",
       "1B3_1    ...                        0.00                     0.00   \n",
       "1B3_2    ...                        0.00                     0.00   \n",
       "1B3_3    ...                        0.00                     0.00   \n",
       "1B3_4    ...                        0.00                     0.00   \n",
       "1B4_0    ...                        0.00                     0.00   \n",
       "1B4_1    ...                        0.00                     0.00   \n",
       "\n",
       "         PRIMER_RIGHT_SEQUENCE PRIMER_RIGHT_TEMPLATE_MISPRIMING  \\\n",
       "1A1_0     CACCCACTACACCCCTCAAC                           7.0000   \n",
       "1A1_1     CCTCCTGTTTCCTCACCCAC                           6.0000   \n",
       "1A1_2     TTGGAGAGGCTAGGAGGGAG                           8.0000   \n",
       "1A1_4     TTCCTCACCCACTACACCCC                           6.0000   \n",
       "1A1_5     CCACTACACCCCTCAACTGT                           7.0000   \n",
       "1A1_6    CCTGTTTCCTCACCCACTACA                           6.0000   \n",
       "1A1_7   GCAGAGAAAGTGACCCCAGAAT                           7.0000   \n",
       "1A1_8   CAGAATTGGAGAGGCTAGGAGG                           7.0000   \n",
       "1A1_9   CTACAAGCAGAGAAAGTGACCC                           7.0000   \n",
       "1A3_0   CACATCATAGGCAGTGCAAAGT                           7.0000   \n",
       "1A3_1   GCTAGTCACATCATAGGCAGTG                           8.0000   \n",
       "1A5_0    ACCTGCATGCTATAAGAGGGC                           8.0000   \n",
       "1A8_0     TGGGACACTGGGACACTACT                          12.0000   \n",
       "1A8_1     GGTCAGTGCCTCTTCTCTGG                           8.0000   \n",
       "1A8_2     GAAGAGCACCCACTGAAGCA                           8.0000   \n",
       "1A8_3      ATCGACCTCCCCTGCCTTT                           8.0000   \n",
       "1A8_4    CCTTTGCAGCTGATTCGAAGA                           8.0000   \n",
       "1A8_5      TCGAAGAGCACCCACTGAA                           7.0000   \n",
       "1A9_0     GACCCCTCCTTGCAATCTGG                           9.0000   \n",
       "1A9_1     CTCCTTGCAATCTGGGTGGG                           7.0000   \n",
       "1A9_2   ACCATCTGTTCACCCTCCTAGT                           8.0000   \n",
       "1A11_0    TCTGTGTCTGACCCAAAGCA                           8.0000   \n",
       "1B1_0     TCTTCTGAAGTAGGCCACGC                           8.0000   \n",
       "1B1_1     GGCCACGCGATCATCACTAT                           7.0000   \n",
       "1B1_2     AAGTAGGCCACGCGATCATC                           6.0000   \n",
       "1B1_3     CTGAAGTAGGCCACGCGATC                           6.0000   \n",
       "1B1_4   AGTGGGCTTCTAAATGTCCTGA                           8.0000   \n",
       "1B3_0     ACTGCGTGGGGTTTTGGTAA                           9.0000   \n",
       "1B3_1   CGTGGGGTTTTGGTAATTTGCT                           8.0000   \n",
       "1B3_2   CTGCGTGGGGTTTTGGTAATTT                           8.0000   \n",
       "1B3_3    AATGTACTCCAAACACTGCGT                           8.0000   \n",
       "1B3_4    TGTCCCAACACAAAGTCACAA                          10.0000   \n",
       "1B4_0    TCCCTTTGCTAACTTTGTGGC                           8.0000   \n",
       "1B4_1   ACCAGACTTTCCCTTTGCTAAC                          10.0000   \n",
       "\n",
       "       PRIMER_RIGHT_TM LEFT_GENOME_HITS POTENTIAL_PRODUCTS  \\\n",
       "1A1_0           59.964              428                  3   \n",
       "1A1_1           59.963              623                  9   \n",
       "1A1_2           59.737              617                  6   \n",
       "1A1_4           60.839              616                  3   \n",
       "1A1_5           58.943              149                  1   \n",
       "1A1_6           59.303              955                  6   \n",
       "1A1_7           60.291              443                  3   \n",
       "1A1_8           59.632             1209                  1   \n",
       "1A1_9           58.670              374                  4   \n",
       "1A3_0           59.249             1359                  6   \n",
       "1A3_1           58.612              132                  3   \n",
       "1A5_0           59.927              295                 22   \n",
       "1A8_0           59.810              340                  1   \n",
       "1A8_1           59.751              842                  7   \n",
       "1A8_2           60.250              776                  5   \n",
       "1A8_3           60.617             1311                  4   \n",
       "1A8_4           58.921              283                  1   \n",
       "1A8_5           58.563              953                  4   \n",
       "1A9_0           60.395              501                  3   \n",
       "1A9_1           60.682              597                  2   \n",
       "1A9_2           60.226              402                  2   \n",
       "1A11_0          59.163              229                  2   \n",
       "1B1_0           59.753              652                  3   \n",
       "1B1_1           60.320              729                  1   \n",
       "1B1_2           60.249              806                  1   \n",
       "1B1_3           60.872              133                  2   \n",
       "1B1_4           59.087              800                  2   \n",
       "1B3_0           60.107              555                  1   \n",
       "1B3_1           59.963              916                  3   \n",
       "1B3_2           59.963              733                  1   \n",
       "1B3_3           58.154              236                  1   \n",
       "1B3_4           58.203              539                  6   \n",
       "1B4_0           59.039              606                  5   \n",
       "1B4_1           58.513              369                  2   \n",
       "\n",
       "                                POTENTIAL_PRODUCT_LENGTHS RIGHT_GENOME_HITS  \\\n",
       "1A1_0                                       646|4624|2654               344   \n",
       "1A1_1         668|4341|2279|4246|2675|4363|3210|2022|4106              1477   \n",
       "1A1_2                        749|4266|1027|2513|2491|1249              1401   \n",
       "1A1_4                                        715|4323|329               606   \n",
       "1A1_5                                                 659               300   \n",
       "1A1_6                           604|4059|2650|3563|78|743               583   \n",
       "1A1_7                                         753|920|101               685   \n",
       "1A1_8                                                 708               363   \n",
       "1A1_9                                   817|3601|749|1815               546   \n",
       "1A3_0                          1416|711|276|594|4270|4225               275   \n",
       "1A3_1                                      1474|4023|4804               326   \n",
       "1A5_0   1358|1355|1355|1358|1355|4206|2825|2779|2113|2...              1801   \n",
       "1A8_0                                                1378               377   \n",
       "1A8_1                  1420|1410|2526|1786|1948|2193|3195              1164   \n",
       "1A8_2                            1908|1399|1670|3669|3672               524   \n",
       "1A8_3                                 1476|4748|2118|4294              1047   \n",
       "1A8_4                                                1358               159   \n",
       "1A8_5                                 1407|2778|4906|3731               647   \n",
       "1A9_0                                       1304|470|4100               335   \n",
       "1A9_1                                            1304|425               273   \n",
       "1A9_2                                           1408|2364               307   \n",
       "1A11_0                                           1175|869               615   \n",
       "1B1_0                                       1157|750|4123               152   \n",
       "1B1_1                                                1149                32   \n",
       "1B1_2                                                1158                28   \n",
       "1B1_3                                           1125|2597                27   \n",
       "1B1_4                                            1190|756               341   \n",
       "1B3_0                                                1159               201   \n",
       "1B3_1                                      1146|1522|3715               439   \n",
       "1B3_2                                                1152               209   \n",
       "1B3_3                                                1147               183   \n",
       "1B3_4                        1094|1491|579|2597|1824|1512               824   \n",
       "1B4_0                            1112|4966|4389|3703|3403               626   \n",
       "1B4_1                                           1178|3693               371   \n",
       "\n",
       "       Selected  \n",
       "1A1_0      True  \n",
       "1A1_1     False  \n",
       "1A1_2     False  \n",
       "1A1_4     False  \n",
       "1A1_5     False  \n",
       "1A1_6     False  \n",
       "1A1_7     False  \n",
       "1A1_8     False  \n",
       "1A1_9     False  \n",
       "1A3_0     False  \n",
       "1A3_1     False  \n",
       "1A5_0     False  \n",
       "1A8_0      True  \n",
       "1A8_1     False  \n",
       "1A8_2     False  \n",
       "1A8_3     False  \n",
       "1A8_4     False  \n",
       "1A8_5     False  \n",
       "1A9_0     False  \n",
       "1A9_1     False  \n",
       "1A9_2     False  \n",
       "1A11_0    False  \n",
       "1B1_0     False  \n",
       "1B1_1      True  \n",
       "1B1_2     False  \n",
       "1B1_3     False  \n",
       "1B1_4     False  \n",
       "1B3_0      True  \n",
       "1B3_1     False  \n",
       "1B3_2     False  \n",
       "1B3_3     False  \n",
       "1B3_4     False  \n",
       "1B4_0      True  \n",
       "1B4_1     False  \n",
       "\n",
       "[34 rows x 32 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_primer_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
